### ISSUE TITLE
Move more information into `type`

### ISSUE THREAD

**Issue by niemela** [Kattis/problem-package-format#131]
I would argue that in everyday language the "type of a problem" would, in addition to pass-fail vs scoring, include properties such as multi-pass, output only, and interactive.

One very informal way of thinking about it could be that if it feels natural to say "It's a X problem", then X is a type. This works with "It's a scoring problem", "It's an interactive problem" (and all the ones I listed above). It also works for "It's a language limited problem" and "It's an attachment having problem", because those are *not* natural, and also IMO, not good examples of type. It does not work for "It's an easy problem"; that *does* feel natural, but I don't think "easy" is a good example of what we should mean by "type". This is clearly not a perfect rule, but maybe a bit helpful? 

In addition to aligning with everyday language, extending type in this way would make type be a better indicator of format support. 

Currently, the following features are not supported by the "ICPC subset":
- 1. Scoring problems
- 2. Multi-pass problems
- 3. Output only problems (I guess?)
- 4. Test data groups
- 5. Language limited problems
- 6. Included files

In addition, the following feature was unsupported by the "ICPC subset" in the past, and would be "reasonable" for some hypothetical system to choose to not support:
- 7. Interactive problems

1-3 & 7 are types.
4 should (IMO) be moved in the ICPC subset
5-6 doesn't feel like types to me (?), but I can see why we might not want it in the ICPC subset. They wouldn't be very hard at all to implement, so it wouldn't be a big ask, but it would probably never be used (?), so it's probably wasted work.

In other words, they type of the problem would be what decides if it's supported (with the expectation of `language` and `include`). This feels good enough to me, and forcing language and include into type as well seems worse.

So, in summary, I suggest that type becomes a list of values:
- at most one of `pass-fail` and `scoring`, defaulting to `pass-fail`
- at most one of `single-pass` and `multi-pass`, defaulting to `single-pass`
- at most one of `input-output` and `output-only`, defaulting to `input-output`
- at most one of `batch` and `interactive`, defaulting to `batch`

Thoughts?

**Comment by thorehusfeldt** [comment]
I have no strong opinion about this, but independent of the taxonomy suggest we move specification formalism more closely towards YAML. This is easier to specify clearly, provides editor support, and is easier to implement (because it is _already parsed_).

For instance
```cue
type?: *"pass-fail" | "scoring" | {
   judgement!: *"verdict" | "score"
  passes?: *"single" | "multiple"
  input?: *true | bool
  interactive?: *false | bool
}
```
This allows 
```yaml
type: pass-fail # this is the default, so this line is redundant
---
type: scoring # for old time's sake; it sets judgement: score
---
type: 
  judgement: score
  passes: multiple
```

I’m not sure all combinations make sense (say, interactive multipass). Also, no strong opinion on `judgement: string`, I’d prefer `scoring: *false | bool` and am open to `multipass: *false | bool` as well.

**Comment by niemela** [comment]
> I [...] suggest we move specification formalism more closely towards YAML

I'm not sure I understand *exactly* what you mean by this?

The data *is* in YAML, how can something that *is* something move more towards that?

If you mean that we should move away from "whitespace separated list of keyword", i.e. a string that is expected to be further parsed, then yes, I agree (and I can see how that is "moving towards YAML").

But are you *also* saying that *list* of keywords are bad, and key-value pairs is "more YAML"? That I don't think I agree with.

> I’m not sure all combinations make sense (say, interactive multipass). Also, no strong opinion on `judgement: string`, I’d prefer `scoring: *false | bool` and am open to `multipass: *false | bool` as well.

All combination except the following make sense:
- `multi-pass` and `output-only`
- `interactive` and `output-only`

**Comment by thorehusfeldt** [comment]
Yes, I merely mean “move away of whitespace-separated lists of keywords” (in particular with various constraints, such as “any of”).  For YAML, maps are the good way of specifying flags, lists are worse (because you cannot easily forbid multiple entries, or even contradictions, like forbidding the list `type: ['scoring', 'scoring', 'pass-fail']`) and strings are even worse (you need really funky regexen to forbid `type: "scoring pass-fail scoring".)

**Comment by Tagl** [comment]
> Is an `output-only` problem a problem where the testcase has no `.in`-file (like `hello`), or one where the solver never reads from standard input? (I think the latter.)

No, it is where you submit the "output" of the program instead of the source code that produces the output. One output file per testcase.

**Comment by thorehusfeldt** [comment]
Got it. To me, `hello` is an output-only problem, so maybe the terminology can be improved. 

It’s about _what_ is validated: A program or a string?  Think about  `validation-target: *"submission" | "output"` or `submission: *"executable" | "output"` or something like that.

**Comment by niemela** [comment]
> Got it. To me, `hello` is an output-only problem, so maybe the terminology can be improved.

Yes, that's a good point.

> It’s about _what_ is validated: A program or a string? Think about `validation-target: *"submission" | "output"` or `submission: *"executable" | "output"` or something like that.

Or rather about what is being submitted (the program or the output) and what is being given (the input or not). In both cases the output is being eventually validated.

**Comment by niemela** [comment]
> I agree that `type` differences between `submissions is a one-shot batch programme that reads input and writes output` and `submission is an interactive progamme` and `submission is a string` are conceptually much larger than the difference between a `pass-fail` and a `scoring` problem. This is indeed worth thinking about.

Huh, interesting. I would definitely have said the other way around (but I guess I can see your point).

For me `interactive` is just a slight generalization of a custom validator, which is just a slight generalization of the default validator which is just a slight generalization of diff. I do think `interactive` makes sense as a `type`, but it was borderline for me.

**Comment by jsannemo** [comment]
...and pass-fail is just a subcase of scoring with an accept score of 1 ;)

**Comment by niemela** [comment]
All that said, we don't really have to perfectly rank the type-ness of different concepts, *but*, is there something in my list above that someone thinks should *not* be a type, or something *not* in the list that *should*?

**Comment by thorehusfeldt** [comment]
I think I’m at:
```cue
type!: *"pass-fail" |  "scoring" |   {
  scoring?: *false | true 
  submission?: *"executable" | "text"
  passes?: *1 | number | "multiple" 
  interactive?: *false | true
}
```
The first two options are shorthands both for legacy and to cover 99% of cases.

We can make the last two fields conditional on `submission` like this:
```cue
type!: *"pass-fail" |  "scoring" |   {
  scoring?: *false | true 
  submission?: *"executable" | "text"
  if submission != _|_ {
    if submission == "executable" {
       passes?: *1 | number | "multiple" 
       interactive?: *false | true
   }
  }
}
```

**Comment by Tagl** [comment]
I'd also like to add my previous comment that single pass is just multipass with n=1, does it really require a new key in problem.yaml or is it enough to just make the output validator output determine whether there is another pass as the specification for it already states?
I personally don't think a new key is required, however, I do see the value in being able to give an error for mismatches in validator behavior and the explicit setting in problem.yaml.

**Comment by niemela** [comment]
> I'd also like to add my previous comment that single pass is just multipass with n=1, does it really require a new key in problem.yaml or is it enough to just make the output validator output determine whether there is another pass as the specification for it already states?

I would like things that systems are likely to not support (*and* that we think is acceptable/reasonable to not support) should be easy to quickly detect. Checking the value of `type` in `problem.yaml` or the presence of the include/ directory is easy and quick. Having to run the validator to see if a `nextpass.in` is produced is *not* easy and quick. I don't think that's good enough.

(Aslo, a sidenote; I think it's even worse than I first thought. You would have to run *all* submissions, because a multi-pass problem would be defined by at least *some* cases being multiple passes, not necessarily all. Right?)

**Comment by Tagl** [comment]
Alright, that is fair.
Yes, there could be a problem with variable number of passes depending on the answer produced by the submission, so some submissions could be run once only for each testcase while others are run some multiple number of times.
